openstack/baremetalinspector/v1/introspection/results.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// IsEmpty returns true if a page contains no Server results.
+func (r IntrospectionPage) IsEmpty() (bool, error) {
+	s, err := ExtractIntrospections(r)
+	return len(s) == 0, err
+}
+
+// NextPageURL uses the response's embedded link reference to navigate to the
+// next page of results.
+func (r IntrospectionPage) NextPageURL() (string, error) {
+	var s struct {
+		Links []gophercloud.Link `json:"introspection_links"`
+	}
+	err := r.ExtractInto(&s)
+	if err != nil {
+		return "", err
+	}
+	return gophercloud.ExtractNextURL(s.Links)
+}
+
+// UnmarshalJSON trie to convert values for started_at and finished_at from the
+// json response into RFC3339 standard. Since Introspection API can remove the
+// Z from the format, if the conversion fails, it falls back to an RFC3339
+// with no Z format supported by gophercloud.
+func (r *Introspection) UnmarshalJSON(b []byte) error {
+	type tmp Introspection
+	var s struct {
+		tmp
+		StartedAt  string `json:"started_at"`
+		FinishedAt string `json:"finished_at"`
+	}
+
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+
+	*r = Introspection(s.tmp)
+
+	if s.StartedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.StartedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.StartedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.StartedAt = t
+	}
+
+	if s.FinishedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.FinishedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.FinishedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.FinishedAt = t
+	}
+
+	return nil
+}
+
+// GetIntrospectionStatusResult is the response from a Get operation.
+// Call its Extract method to interpret it as an Introspection.
+type GetIntrospectionStatusResult struct {
+	introspectionResult
+}
+
+// StartResult ...
+type StartResult struct {
+	gophercloud.ErrResult
+}
+
+// AbortResult ...
+type AbortResult struct {
+	gophercloud.ErrResult
+}
+
+// Data represents introspection data collected.
+// The format and contents of the stored data depends on the ramdisk used
+// and plugins enabled both in the ramdisk and in inspector itself.
+// This structure has been provided for basic compatibility but it
+// will need extensions
+type Data struct {
+	CPUArch       string                 `json:"cpu_arch"`
+	MACs          []string               `json:"macs"`
+	RootDisk      RootDiskType           `json:"root_disk"`
+	Interfaces    map[string]interface{} `json:"interfaces"`
+	CPUs          int                    `json:"cpus"`
+	BootInterface string                 `json:"boot_interface"`
+	MemoryMB      int                    `json:"memory_mb"`
+	IPMIAddress   string                 `json:"ipmi_address"`
+	Inventory     InventoryType          `json:"inventory"`
+	Error         string                 `json:"error"`
+	LocalGB       int                    `json:"local_gb"`
+	AllInterfaces map[string]interface{} `json:"all_interfaces"`
```

Comment:
Speaking of poorly defined interfaces, the `extra_hardware` plugin defines the `extra` key, which is a 3 level nested dict: https://github.com/openstack/ironic-inspector/blob/master/ironic_inspector/plugins/extra_hardware.py#L66

Reply:
My understanding is that it will not cause any issue if it's not present in the structure.
Of course it will be interpreted and not formally defined.